rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
+
+GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.
+
+Focusing on containerisation will detract from work to improve other parts of the platform.
+
+GOV.UK has a significant amount of infrastructure outside of the applications themselves. In moving to a container based hosting platform, it’s possible that some existing behaviour may be missed, resulting in bugs.
+### Risks of not doing this
+
+GOV.UK reduces its options of how to mitigate Trusty reaching the end of support so the only option left is upgrading Ubuntu. This is potentially wasted work as once the upgrade is done GOV.UK will want to consider if running VMs is best for GOV.UK.
+
+GOV.UK will also have to tackle a number of difficult tooling upgrades (Puppet, in particular) if GOV.UK wants to return to a situation where GOV.UK does not use unsupported dependencies.
+
+GOV.UK will be harder for the rest of GDS to support, because its infrastructure will be significantly different. Almost all other service teams in GDS are using a container based hosting environment of some kind (mostly [CloudFoundry](https://www.cloudfoundry.org) or AWS ECS).
+### Follow up work
+
+GOV.UK will commit to doing these things:
+
+- Investigate the practicalities involved in using different containerised runtimes to host GOV.UK (for example, how difficult will it be to run a hybrid containerised / non-containerised environment in [GOV.UK PaaS](https://www.cloud.service.gov.uk) vs. say [AWS ECS](https://aws.amazon.com/ecs/))
+- Understand how containerising will affect our CI/CD, metrics & alerting, and support models.
```

Comment:
We don't do CD as far as I know (regardless of whether this means Continuous Delivery or Continuous Deployment). Are you saying there's a commitment to understand how Continuous D(something) would be done?

Reply:
```suggestion
- Understand how containerising will affect our build and deployment process, observability, alerting, and support models.
```

Reply:
> Are you saying there's a commitment to understand how Continuous D(something) would be done?

I don't think so.

We certainly need to make sure that we can deploy to whatever platform we end up with, and it needs to *at least* not be worse than the current situation.

There's an assumption in this RFC that a container-based platform would make CD easier to implement.

I think the actual decision on whether we want to do CD as part of this work, or whether we want to do it at all, still needs making (and needs to include technical, organisational and human aspects).

Reply:
Turns out we've already got [an RFC for CD accepted](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-070-path-towards-continuous-deployment-cd.md). But this RFC isn't trying to say any more on that subject. I've cleared up the wording not to mention CD in 4d3e92f47b46b679d565891744a7b1361aa43a78.